File: C++/reshape_the_matrix.cpp
#include <iostream>
#include "./reshape_the_matrix.h"
#include "./utility.h"

using namespace std;

vector<vector<int> > Solution::matrixReshape(vector<vector<int> >& nums, int r, int c) {
    // error handing
    int origin_rows = nums.size();
    int origin_cols =  origin_rows > 0 ? nums[0].size(): 0;
    int product = origin_cols * origin_rows;
    if (product == 0 || product != (r * c)) {
        cout << "sxxxxx";
        return nums;
    }

    vector<vector<int> > output;
    int origin_row_idx = 0, origin_col_idx = 0;
    // O(r*c)
    for (int i = 0; i < r; ++i) {
        vector<int> this_row;
        for (int j = 0; j < c; ++j) {
            this_row.push_back(nums[origin_row_idx][origin_col_idx]);
            origin_col_idx++;
            // carriage return
            if (origin_col_idx == origin_cols) {
                origin_col_idx = 0;
                origin_row_idx += 1;
            }
        }
        output.push_back(this_row);
    }

    return output;
}

int main() {
    vector<vector<int> > A;
    
    vector<int> r1;
    r1.push_back(1);
    r1.push_back(2);
    vector<int> r2;
    r2.push_back(3);
    r2.push_back(4);
    
    A.push_back(r1);
    A.push_back(r2);

    Solution* sol = new Solution();
    vector<vector<int> > output = sol->matrixReshape(A, 1, 4);
    delete sol;

    Utility::printVector(output[0]);

    return 0;
}

File: Python2/reshape_the_matrix.py
class Solution(object):
    origin_rows = 0
    origin_cols = 0

    def matrixReshape(self, nums, r, c):
        """
        :type nums: List[List[int]]
        :type r: int
        :type c: int
        :rtype: List[List[int]]
        """
        self.origin_rows = len(nums)
        self.origin_cols = len(nums[0]) if self.origin_rows > 0 else 0
        if self.origin_cols*self.origin_rows == 0 or (self.origin_rows*self.origin_cols) != (r*c):
            return nums

        return self.optimized_way(nums, r, c)

    def optimized_way(self, nums, r, c):
        lst = [y for x in nums for y in x]
        # output = []
        # for i in range(r):
        #     output.append(lst[i*c:i*c+c])
        #return output
        return [lst[i*c:i*c+c] for i in range(r)]

    def easy_way(self, nums, r, c):
        output = []
        origin_row_idx = origin_col_idx = 0
        for _ in range(r):
            this_row = []
            for _ in range(c):
                print origin_row_idx, origin_col_idx
                this_row.append(nums[origin_row_idx][origin_col_idx])
                origin_col_idx += 1
                if origin_col_idx == self.origin_cols:
                    origin_row_idx += 1
                    origin_col_idx = 0
            output.append(this_row)
        return output


if __name__ == '__main__':
    nums = [[1, 2], [3, 4]]
    r = 1
    c = 4
    sol = Solution()
    print sol.matrixReshape(nums, r, c)
